File: Python/Arrays/L_FindDuplicate.py
from typing import List


def findDuplicate(nums: List[int]) -> int:
    h = t =nums[0]
    while True:
        t =nums[t]
        h = nums[nums[h]]
        if t == h:
            break
    t = nums[0]
    while t != h:
        t = nums[t]
        h = nums[h]
    return t

def findDuplicate1(nums: List[int]) -> int:
        slow,fast=0,0
        while True:
            slow=nums[slow]
            fast=nums[nums[fast]]
            if slow==fast:
                break

        slow2=0
        while True:
            slow=nums[slow]
            slow2=nums[slow2]
            if slow==slow2:
                return slow   

File: cpp/Array/L_findDuplicate.cpp
#include <bits/stdc++.h>
using namespace std;
int findDuplicate(vector<int> &nums)
{
    int slow = 0, fast = 0;
    do
    {
        slow = nums[slow];
        fast = nums[nums[fast]];
    } while (slow != fast);
    int slow1 = 0;
    do
    {
        slow = nums[slow];
        slow1 = nums[slow1];
    } while (slow != slow1);
    return slow;
}

int findDuplicate(std::vector<int> &nums)
{
    int left = 1;
    int right = nums.size() - 1;

    while (left < right)
    {
        int mid = left + (right - left) / 2;
        int count = 0;

        // Count the numbers less than or equal to mid
        for (int num : nums)
        {
            if (num <= mid)
            {
                count++;
            }
        }

        // If count is greater than mid, the duplicate lies in the left half
        if (count > mid)
        {
            right = mid;
        }
        else
        { // Otherwise, it lies in the right half
            left = mid + 1;
        }
    }

    return left;
}
int main()
{
}